### PR TITLE
NF: replace `tree` with python implementation

### PR DESCRIPTION
**Pull request by aqw**
This implementation mimics the original behavior, while fixing two bugs:
    
- `onyo tree` now prints from CWD (as documented) rather than repository root. The `onyo_tree` function still defaults to repository root.
 - argument dirs are now printed exactly as the user specified, rather than always absolute
    
By using pure Python, we gain flexibility to add Onyo-specific features (perhaps asset highlighting, etc) and also drop a dependency (albeit a common one).
    
We never allowed passing flags to `tree`, so there really is no loss in functionality.
    
I did some primitive performance tests.
    

```
32 directories, 42 files:
-            tree  0.00s user 0.00s system 87% cpu 0.005 total
- (old) onyo tree  0.09s user 0.01s system 99% cpu 0.094 total
- (new) onyo tree  0.09s user 0.01s system 99% cpu 0.101 total
```

```
1844 directories, 9706 files
-            tree  0.06s user 0.03s system 99%  cpu 0.090 total
- (old) onyo tree  0.12s user 0.05s system 106% cpu 0.164 total
- (new) onyo tree  1.21s user 0.08s system 100% cpu 1.289 total
```
    
In the "normal" use case, where someone prints a tree of a size that a human can reasonably read, the performance hit is negligible.

However, large trees do suffer from a noticeable performance impact. This can be dropped to 0.43s if all output is printed at once.
```
1844 directories, 9706 files
- onyo tree  0.38s user 0.07s system 102% cpu 0.435 total
 ```

### CODE REVIEW

**Review by TobiasKadelka — COMMENTED** [review]
All in all looks good, and I really like the overall change, using a python function instead of calling `tree`!

I have remarked some small points that I think are worth considering.

File: onyo/lib/commands.py [hunk 1]
```diff
@@ -1042,36 +1042,82 @@ def onyo_set(inventory: Inventory,
...
     Parameters
     ----------
     inventory: Inventory
         The inventory in which the directories to display are located.
 
-    paths: list of Path
-        The paths to directories for which to print the directory tree.
-        If no path is specified, `onyo_tree(inventory)` prints the
-        directory tree for the root of the inventory.
+    dirs: list tuple (str, Path)
```
[the remark this answers is not in the document]
Comment (aqw) [reply onyo/lib/commands.py:1053]: I was unsure, so I copied for elsewhere in the codebase. Happy to adjust this though.

[the remark this answers is not in the document]
Comment (aqw) [reply onyo/lib/commands.py:1053]: Ahh, damn. Now I see you meant the "of". I agree. Will fix.

File: onyo/lib/commands.py [hunk 2]
```diff
@@ -1042,36 +1042,82 @@ def onyo_set(inventory: Inventory,
...
 
     Raises
     ------
     ValueError
         If paths are invalid.
     """
     # sanitize the paths
-    dirs = paths if paths else [inventory.root]
-    non_inventory_dirs = [str(p) for p in dirs if not inventory.repo.is_inventory_dir(p)]
+    dirs = dirs if dirs else [('.', inventory.root)]
```
[the remark this answers is not in the document]
Comment (aqw) [reply onyo/lib/commands.py:1070]: Yes. I felt that if we're implicitly operating from inventory root, then that's the scope of our world view. Precisely because it's implicit, it means the user doesn't imagine the outside world.

Including the FS path leading to the repo root would seem to run counter to that idea.

File: onyo/lib/commands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bpoldrack) [reply onyo/lib/commands.py:1053]: Not exactly me having strong opinions, but numpy. ;-)

However, for nested declarations numpy isn't very well defined IMO. I couldn't find any precedence rule for `or` vs `of` for example. So I'd just go for only specifying first level here:

```suggestion
    dirs: list of tuple
```

File: onyo/lib/commands.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bpoldrack) [reply onyo/lib/commands.py:1070]: Good catch, that this is a debatable spot. I wouldn't mind much, though:

1. Both - `'.'` and absolute path - are defensible options from the python interface POV, IMO. You know what `Inventory` object you are calling this on.
2. I don't really see a python interface usecase for this type of command to begin with, other than more sophisticated scripts that are meant to be used via CLI themselves. In which case you wouldn't really use the default anyway.

In short: I think it's fine either way.

File: onyo/lib/commands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aqw) [reply onyo/lib/commands.py]: Thanks. :-)

Fixed and pushed.

File: onyo/lib/commands.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aqw) [reply onyo/lib/commands.py:1070]: > I don't really see a python interface usecase for this type of command to begin with

I agree, and even less-so for using the python interface to generically list a tree of repo root without specifying the directory.

I actually think the right thing to do here is to remove the optional nature of the `dirs` argument. It removes the ambiguity of this situation, and allows me to remove the silencing of Pyre.

[the remark this answers is not in the document]
Comment (aqw) [reply onyo/lib/commands.py]: I have made `dirs` required, and pushed.
